### PR TITLE
tech debt: simplify access log code by adding a ALL_RESPONSE_STRING_FLAGS

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
Remove a bunch of duplicated code in the access log. Adding new response flag will be easier. 

Signed-off-by: Yuchen Dai <silentdai@gmail.com>

Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by lambdai** [comment]
https://storage.googleapis.com/envoy-pr/88caf8e/coverage/source/common/stream_info/index.html

The coverage is lower because I rewrite the verbose code :sob:

File: source/common/stream_info/utility.cc [hunk 1]
```diff
@@ -32,6 +32,41 @@ const std::string ResponseFlagUtils::DURATION_TIMEOUT = "DT";
 const std::string ResponseFlagUtils::UPSTREAM_PROTOCOL_ERROR = "UPE";
 const std::string ResponseFlagUtils::NO_CLUSTER_FOUND = "NC";
 
+const std::vector<std::pair<std::string, ResponseFlag>>
```
Comment (jmarantz) [inline source/common/stream_info/utility.cc:35]: non statically constructed non-PODs please.

Actually the above existing std::string declarations also violate this but I think you can replace them with absl::string_view which are safe to statically construct from string literals. See https://abseil.io/tips/140

I am not actually sure if it's kosher to make a static array of pairs. I suspect it's OK, but I'd at least follow the pattern in https://stackoverflow.com/questions/55099871/iinitializing-a-constexpr-stdarray-of-pairs and
 * use std::array rather than std::vector
 *  declare it constexpr, not const

Comment (jmarantz) [reply source/common/stream_info/utility.cc:35]: https://quuxplusone.github.io/blog/2018/06/26/cost-of-static-lifetime-constructors/ also makes interesting reading.

Comment (lambdai) [reply source/common/stream_info/utility.cc:35]: neither string nor vector has constexpr constructor. Fortunately, pair is not a blocker.

array might resolve the vector, but string_view is not always a replacement since we need `const string&` to the string.
Let me give it another try to see if string_view with a limit amount `const string&`  to `string_view` works.

If it's too large I will use static strings variable within a function

Comment (jmarantz) [reply source/common/stream_info/utility.cc:35]: Yeah plus the current state is not IMO acceptable. We should not have static std::strings being constructed. in Envoy.

You may need to also change some existing APIs to take absl::string_view to make this all work.

Comment (jmarantz) [reply source/common/stream_info/utility.cc:35]: I'll just add also that absl::string_view was not available at the time some of the older Envoy code was written, so there are still some APIs that require `const std::string&`. But I think it's not too hard to get alterations of that form through review.

File: source/common/stream_info/utility.cc [hunk 2]
```diff
@@ -5,73 +5,48 @@
...
-const std::string ResponseFlagUtils::STREAM_IDLE_TIMEOUT = "SI";
-const std::string ResponseFlagUtils::INVALID_ENVOY_REQUEST_HEADERS = "IH";
-const std::string ResponseFlagUtils::DOWNSTREAM_PROTOCOL_ERROR = "DPE";
-const std::string ResponseFlagUtils::UPSTREAM_MAX_STREAM_DURATION_REACHED = "UMSDR";
-const std::string ResponseFlagUtils::RESPONSE_FROM_CACHE_FILTER = "RFCF";
-const std::string ResponseFlagUtils::NO_FILTER_CONFIG_FOUND = "NFCF";
-const std::string ResponseFlagUtils::DURATION_TIMEOUT = "DT";
-const std::string ResponseFlagUtils::UPSTREAM_PROTOCOL_ERROR = "UPE";
-const std::string ResponseFlagUtils::NO_CLUSTER_FOUND = "NC";
+// const std::array<std::pair<absl::string_view, ResponseFlag>>
```
Comment (jmarantz) [inline source/common/stream_info/utility.cc:8]: cool; can this be constexpr?

Comment (lambdai) [reply source/common/stream_info/utility.cc]: I guess the feedback should be "delete the stale comment!!!" :)

 It's constexpr in utility.h

File: test/per_file_coverage.sh [hunk 3]
```diff
@@ -18,6 +18,7 @@ declare -a KNOWN_LOW_COVERAGE=(
 "source/common/protobuf:94.6"
 "source/common/signal:84.5" # Death tests don't report LCOV
 "source/common/singleton:95.1"
+"source/common/stream_info:96.4"
```
Comment (jmarantz) [inline test/per_file_coverage.sh:21]: Why is this needed? Is this because you brought down the line-count making the coverage % worse? Any chance we could add coverage for missing lines?

**Review by jmarantz — COMMENTED** [review]
looks great; just a few small comments. Thank you!

File: test/per_file_coverage.sh (hunk 3 above)
Comment (lambdai) [reply test/per_file_coverage.sh:20]: Yes, this is unfortunate.

> https://storage.googleapis.com/envoy-pr/88caf8e/coverage/source/common/stream_info/index.html
> 
> The coverage is lower because I rewrite the verbose code

That could be done in separate PRs. It's likely that code from elsewhere will be touched

Comment (jmarantz) [reply test/per_file_coverage.sh:20]: @alyssawilk for judgement on this point. @lambdai would you need to change prod code or just add unit tests.

Comment (alyssawilk) [reply test/per_file_coverage.sh:20]: This is a legit reason to add an exception but given @lambdai is a frequent contributor I'd ask we just add unit tests here, even if they're for unrelated code.

testing upstreamSslConnection and serializeAsProto should be super easy and also get us back over the bar.

Comment (lambdai) [reply test/per_file_coverage.sh:20]: Thank you! I am adding now

Comment (alyssawilk) [reply test/per_file_coverage.sh:20]: Awesome!  Thanks so much for being willing to :-)

Comment (lambdai) [reply test/per_file_coverage.sh]: The coverage is over the required percentage now. I have reverted the override coverage value. Can you take another look?

File: source/common/stream_info/utility.cc [hunk 4]
```diff
@@ -5,180 +5,44 @@
...
-const std::string ResponseFlagUtils::NO_CLUSTER_FOUND = "NC";
-
-void ResponseFlagUtils::appendString(std::string& result, const std::string& append) {
+void ResponseFlagUtils::appendString(std::string& result, absl::string_view append) {
   if (result.empty()) {
     result = append;
   } else {
-    result += "," + append;
+    result += ",";
+    result += append;
```
Comment (jmarantz) [inline source/common/stream_info/utility.cc:13]: perf nit: `absl::StrCat(&result, ",", append);` will right-size the buffer in result, where as separately appending twice may cause 2 resizes.

File: source/common/stream_info/utility.cc [hunk 5]
```diff
@@ -5,180 +5,44 @@
...
-    appendString(result, DELAY_INJECTED);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::FaultInjected)) {
-    appendString(result, FAULT_INJECTED);
-  }
-
-  if (stream_info.hasResponseFlag(ResponseFlag::RateLimited)) {
-    appendString(result, RATE_LIMITED);
+  for (const auto& [flag_string, flag] : ALL_RESPONSE_STRING_FLAGS) {
```
Comment (jmarantz) [inline source/common/stream_info/utility.cc:20]: actually the above comment is superfluous. What you really want to do is make a temp `vector<string_view>` and then join that:
```
std::vector<absl::string_view> flag_strings;
for (const auto& [flag_string, flag] : ALL_RESPONSE_STRING_FLAGS) {
  if (stream_info.hasResponseFlag(flag)) {
    flag_strings.push_back(flag_string);
  }
}
if (flag_strings.empty()) {
  return NONE;
}
return absl::StrJoin(flag_strings, ",");
```
This will result in a lot less string resizing and appending.

TBH I'm not sure if this is called enough to warrant optimization but the optimized version is not too complex.

**Review by jmarantz — COMMENTED** [review]
just a minor perf nit

File: source/common/stream_info/utility.cc (hunk 4 above)
Comment (lambdai) [reply source/common/stream_info/utility.cc]: adopt the below tech

File: source/common/stream_info/utility.cc (hunk 5 above)
Comment (lambdai) [reply source/common/stream_info/utility.cc:13]: That's awesome! I replace the std::vector by InlinedVector and copy-paste each character below.

File: source/common/stream_info/utility.cc [hunk 6]
```diff
@@ -1,29 +1,24 @@
...
-    result += ",";
-    result += append;
-  }
-}
-
 const std::string ResponseFlagUtils::toShortString(const StreamInfo& stream_info) {
-  std::string result;
-
+  absl::InlinedVector<absl::string_view,
+                      std::numeric_limits<std::underlying_type_t<ResponseFlag>>::digits>
```
Comment (jmarantz) [inline source/common/stream_info/utility.cc:11]: what is this number, btw? 

No idea what the impact of having (say) 100 here when you only need 3, but if that number is like 100000 you might blow out the stack.

Comment (lambdai) [reply source/common/stream_info/utility.cc:11]: The underlying type of `enum class ResponseFlag` is int at my platform and the ::digit number is 31

Maybe some platform could use 64 or 63 but no more than that.
The stack should be safe IMO

**Review by jmarantz — DISMISSED** [review]
Great, thanks!

Might need to merge main to try to pass CI?

@envoyproxy/senior-maintainers

**Comment by alyssawilk** [comment]
CI is broken right now but
/assign @envoyproxy/senior-maintainers

**Comment by alyssawilk** [comment]
boo.  @htuch @itayd did I do the assign incorrectly or is it opt-in per group?

**Comment by alyssawilk** [comment]
hm, I think I have the right command this time and it's still not working.  https://github.com/envoyproxy/envoy/pull/15594 is merged - do we need to update anything else to get it to work?

**Comment by htuch** [comment]
@itayd any idea it didn't pick up when @alyssawilk tried? It looks like no errors on #repokitteh and no indication of RK observing the comment. @alyssawilk can you try out the command in some other PRs?

File: source/common/stream_info/utility.cc (hunk 6 above)
Comment (mattklein123) [reply source/common/stream_info/utility.cc:11]: Can you add some comments? I don't immediately understand what this does. It seems like it would be more clear to just pick an arbitrary number like 1, since in most cases I think there will be a single flag.

File: source/common/stream_info/utility.cc [hunk 7]
```diff
@@ -1,184 +1,43 @@
...
+    }
+  }
+  if (flag_strings.empty()) {
+    return std::string(NONE);
+  }
+  return absl::StrJoin(flag_strings, ",");
+}
 
+absl::flat_hash_map<std::string, ResponseFlag> ResponseFlagUtils::getFlagMap() {
   static_assert(ResponseFlag::LastFlag == 0x1000000, "A flag has been added. Fix this code.");
```
Comment (mattklein123) [inline source/common/stream_info/utility.cc:25]: Is this still needed? Same elsewhere?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks nice cleanup! A few small comments.

/wait

File: source/common/stream_info/utility.cc (hunk 6 above)
Comment (lambdai) [reply source/common/stream_info/utility.cc:11]: Yeah, usually 1 but InlinedVector is free. I can add the comment

Comment (mattklein123) [reply source/common/stream_info/utility.cc:12]: Sorry for being a readability pedant, but I just don't think using all of this templating stuff adds any value here? Can you just use 1, or 16 or whatever? This seems like a roundabout way to pick a small number.

/wait

**Comment by itayd** [comment]
@htuch seems like a transient github error? from the trace it returned 404, but after that was ok.
```
github_call(
  method="GET",
  path="repos/envoyproxy/envoy/assignees/envoyproxy/senior-maintainers",
  body=null,
  success_codes=[204,404],
  accept=null,
) -> (
  {"body":"","json":null,"links":{},"status":"404 Not Found","status_code":404}, (
  null, 
)
```

File: source/common/stream_info/utility.cc (hunk 6 above)
Comment (lambdai) [reply source/common/stream_info/utility.cc:12]: I like 4+. Then 16 since it's in your provided choices :)

**Comment by lambdai** [comment]
Do I need to merge main?

File: source/common/stream_info/utility.cc [hunk 8]
```diff
@@ -1,184 +1,44 @@
 #include "common/stream_info/utility.h"
 
 #include <string>
+#include <type_traits>
```
Comment (mattklein123) [inline source/common/stream_info/utility.cc:4]: nit: I think you can remove this now

**Review by mattklein123 — COMMENTED** [review]
Thanks LGTM with small comment.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

File: source/common/stream_info/utility.cc (hunk 8 above)
Comment (lambdai) [reply source/common/stream_info/utility.cc]: fixed
